Feature: Change speed of calendar progress

2TallTyler opened the pull request:
## Motivation / Problem

In the [daylength](https://github.com/OpenTTD/OpenTTD/discussions/8397#discussioncomment-224382) discussion, @andythenorth gives motivations for players to configure how fast technology progresses in OpenTTD:

> Play 1960 trains forever!
> 
> For many players, the game seems to progress quickly. A new generation of vehicles appears, so we upgrade and tune our networks. Then just as we finish, boom! A new generation of vehicles appears.
> 
> On busy or large maps it can be hard to keep up.
> 
> Slower progression of vehicles would make a more relaxing games for some players.

Additional usecases might be:

* A GS freezes technology progress and advances it when the player completes transport goals ("research" in many games).
* Early-year games such as the American Wild West can actually reach a late-game network size and bank account without cheating time and money — without this, by the time a large network can be built, the era is long past.
* Industry sets can be designed for frozen progression without regard to supporting realistic technological progression, and can offer industry chains which were phased out long ago. Imagine sending ships to hunt whales, delivering coal to gasworks for town gas production, and harvesting ice from frozen lakes to ship around the world. Currently, the NewGRF author can either choose to force these industries to close, frustrating the player, or keep them around forever, which feels silly.

## Description

This implements @nielsmh's [approach described](https://github.com/OpenTTD/OpenTTD/pull/9789#issuecomment-1013852571) in my previous attempt to implement this (#9789).

Date/time handling is separated into two systems:
#### Calendar time for technology and world appearance
- [x] Vehicle introduction and obsolescence
- [x] Available airports, stations, objects, etc.
- [x] NewGRF variables which can be used to determine visual styles or behavior based on the time of year
- [x] Snow line movement

#### Economy time
- [x] Industry and house production/consumption, etc.
- [x] Town growth
- [x] Industry production changes, closure, and spawning
- [x] NewGRF variables for things like FIRS production boosts
- [x] AI/GS API additions for things like town growth scripts
- [x] Running and maintenance costs, loan interest, etc.
- [x] Vehicle and group profit statistics
- [x] Finance graphs
- [x] Cargo payment rates graph
- [x] Local authority reputation, station ratings, etc.
- [x] Vehicle service interval and time since last service
- [x] Linkgraph tooltips and update interval setting
- [x] Autosave frequency and other settings
- [ ] Network game duration in server list (replacing "Years") - @TrueBrain will help with this before merging.

Calendar time is kept in a common Gregorian calendar, with 365 days in a year, leap years, variable length months, and so on.

Economy time is kept in (approximate) real time. There are logical days with a duration of 1.998 seconds, logical months with a duration of 1 minute, and logical years with a duration of slightly more than 12 minutes.

Internally a calendar with 12 months of 30 days is observed. There are no leap periods. 

In the user interface, all elements that run on economy time are presented in real time units:
#### Seconds (multiples of 2), for things which were originally daily
- [x] Timetables
#### Minutes, for things which were originally monthly:
- [x] Industry production per minute
- [x] Passengers last minute
- [x] Vehicle service intervals and time since last service
- [x] Subsidies
 #### Economic periods (sometimes described as simply 12 minutes) for things which were originally yearly:
- [x] Running costs
- [x] Company finances
- [x] Vehicle and group profit statistics
- [x] Company performance rating (league table) 

Internally most events will still be referenced as daily/monthly/yearly.

## Limitations

This changes MILLISECONDS_PER_TICK from 30ms to 27ms to make an in-game day last 1.99 seconds instead of 2.2 seconds, making real-world time feasible. This actually matches TTD and closes #10205. 😉 

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

nielsmh committed:
Change: Make tick length 27 milliseconds

This makes a month last about 60 seconds, allowing the use of real-time units in game.

2TallTyler committed:
Add: Track economy date in additon to calendar date

2TallTyler committed:
Change: Split daily/monthly/yearly loops into economy and calendar loops

2TallTyler committed:
Add: Saveload conversions for economy dates

2TallTyler committed:
Change: Use minutes for vehicle service interval and last service date

2TallTyler committed:
Feature: Setting to control calendar date progress speed

2TallTyler committed:
Change: Use minutes for various strings

James103 commented on src/date_type.h:
```diff
@@ -22,14 +22,18 @@ typedef uint8  Day;   ///< Type for the day of the month, note: 1 based, first d
 static const int MONTHS_IN_YEAR    =  12; ///< months per year
 
+static const int DAYS_IN_ECONOMY_MONTH  = 30;  ///< economy days per econony month -- at 2 seconds a day this equals 1 minute per month
+static const int MONTHS_IN_ECONOMY_YEAR = 12; ///< months in economy economy year
```
```suggestion
static const int DAYS_IN_ECONOMY_MONTH  = 30; ///< economy days per econony month -- at 2 seconds a day, this equals 1 minute per month
static const int MONTHS_IN_ECONOMY_YEAR = 12; ///< months in economy economy year
```

James103 commented on src/gfx_type.h:
```diff
@@ -313,8 +313,12 @@ enum SpriteType : byte {
+/**
+ * The number of milliseconds per game tick.
+ * The value 27 together with a day length of 74 ticks makes one day 1998 milliseconds, almost exactly 2 seconds.
+ * With a 2 second day, one standard month is 1 minute, and one standard year slightly over 12 minutes.
```
```suggestion
 * With a 2 second day, one standard month is 1 minute, and one standard year is slightly over 12 minutes.
```

James103 commented on src/lang/english.txt:
```diff
@@ -3494,10 +3506,10 @@ STR_LOCAL_AUTHORITY_ACTION_BRIBE                                :Bribe the local
 STR_LOCAL_AUTHORITY_ACTION_TOOLTIP_STATUE_OF_COMPANY            :{PUSH_COLOUR}{YELLOW}Build a statue in honour of your company.{}Provides a permanent boost to station rating in this town.{}{POP_COLOUR}Cost: {CURRENCY_LONG}
 STR_LOCAL_AUTHORITY_ACTION_TOOLTIP_NEW_BUILDINGS                :{PUSH_COLOUR}{YELLOW}Fund the construction of new buildings in the town.{}Provides a temporary boost to town growth in this town.{}{POP_COLOUR}Cost: {CURRENCY_LONG}
-STR_LOCAL_AUTHORITY_ACTION_TOOLTIP_EXCLUSIVE_TRANSPORT          :{PUSH_COLOUR}{YELLOW}Buy 1 year's exclusive transport rights in town.{}Town authority will not allow passengers and cargo to use your competitors' stations.{}{POP_COLOUR}Cost: {CURRENCY_LONG}
+STR_LOCAL_AUTHORITY_ACTION_TOOLTIP_EXCLUSIVE_TRANSPORT          :{PUSH_COLOUR}{YELLOW}Buy 12 minute's exclusive transport rights in town.{}Town authority will not allow passengers and cargo to use your competitors' stations.{}{POP_COLOUR}Cost: {CURRENCY_LONG}
```
```suggestion
STR_LOCAL_AUTHORITY_ACTION_TOOLTIP_EXCLUSIVE_TRANSPORT          :{PUSH_COLOUR}{YELLOW}Buy 12 minutes of exclusive transport rights in town.{}Town authority will not allow passengers and cargo to use your competitors' stations.{}{POP_COLOUR}Cost: {CURRENCY_LONG}
```

Alternatively, if the apostrophe is to be kept, that part could instead say "Buy 12 minutes' exclusive transport rights ..."

James103 commented on src/saveload/saveload.h:
```diff
@@ -344,6 +344,8 @@ enum SaveLoadVersion : uint16 {
 	SLV_LAST_LOADING_TICK,                  ///< 301  PR#9693 Store tick of last loading for vehicles.
 	SLV_MULTITRACK_LEVEL_CROSSINGS,         ///< 302  PR#9931 Multi-track level crossings.
 
+	SLV_NOCALENDAR,                         ///< 303  PR#xxxx Separate calendar time and economy time.
```
```suggestion
	SLV_NOCALENDAR,                         ///< 303  PR#10322 Separate calendar time and economy time.
```

James103 reviewed (commented):
I found some minor items whose grammar could be improved.

2TallTyler committed:
Codechange: Don't assert when using date cheat with slow calendar progress

2TallTyler committed:
Change: Use seconds for timetables

2TallTyler committed:
Remove: Unused date selection GUI

2TallTyler committed:
Change: Use 12-minute economic periods for company finance window

2TallTyler committed:
Update: Be nice to translators

2TallTyler committed:
Change: Use seconds for AI start delay

2TallTyler committed:
Change: Use seconds for town growth rate

2TallTyler committed:
Change: Use seconds and economy dates for Linkgraphs

2TallTyler committed:
Change: Use tiles/sec for game unit velocity

2TallTyler committed:
Change: Use seconds on cargo payment graph time axis

2TallTyler committed:
Change: Use economic periods for vehicle and group profits

2TallTyler committed:
Change: Use economic periods for various settings

2TallTyler committed:
Change: Use economic periods for various cost strings

2TallTyler committed:
Change: Use minutes for subsidies

2TallTyler committed:
Change: Use economy dates for industry production changes and closures

2TallTyler committed:
Change: Use economy dates for economy statistics and interest payments

2TallTyler committed:
Change: Use economy years for minimum company age to trade shares

2TallTyler committed:
Change: Use economy dates to clean up old news items

2TallTyler committed:
Change: Use periods for vehicle age

2TallTyler committed:
Change: Use minutes for company graphs x-axis

2TallTyler committed:
Add: NewGRF global variables for economy date and calendar speed progress

2TallTyler committed:
Add: AI/GS API GetEconomyDate() method

Note: Getting and setting the calendar progress speed is already possible using GSGameSettings
